src/bytes_mut.rs

```diff
@@ -1246,76 +1094,107 @@ impl Shared {
-        if width == MIN_ORIGINAL_CAPACITY_WIDTH + 1 {
-            assert_eq!(original_capacity_to_repr(cap - 24), expected - 1);
-            assert_eq!(original_capacity_to_repr(cap + 76), expected);
-        } else if width == MIN_ORIGINAL_CAPACITY_WIDTH + 2 {
-            assert_eq!(original_capacity_to_repr(cap - 1), expected - 1);
-            assert_eq!(original_capacity_to_repr(cap - 48), expected - 1);
+        if (*self_ptr).refcount.fetch_sub(1, Ordering::Release) != 1 {
+            return;
         }
-    }
-}
 
-#[test]
-fn test_original_capacity_from_repr() {
-    assert_eq!(0, original_capacity_from_repr(0));
+        // The shared state should be released
 
-    let min_cap = 1 << MIN_ORIGINAL_CAPACITY_WIDTH;
+        // This fence is needed to prevent reordering of use of the data and
+        // deletion of the data.  Because it is marked `Release`, the decreasing
+        // of the reference count synchronizes with this `Acquire` fence. This
+        // means that use of the data happens before decreasing the reference
+        // count, which happens before this fence, which happens before the
+        // deletion of the data.
+        //
+        // As explained in the [Boost documentation][1],
+        //
+        // > It is important to enforce any possible access to the object in one
+        // > thread (through an existing reference) to *happen before* deleting
+        // > the object in a different thread. This is achieved by a "release"
+        // > operation after dropping a reference (any access to the object
+        // > through this reference must obviously happened before), and an
+        // > "acquire" operation before deleting the object.
+        //
+        // [1]: (www.boost.org/doc/libs/1_55_0/doc/html/atomic/usage_examples.html)
+        atomic::fence(Ordering::Acquire);
+
+        alloc::alloc::dealloc(
+            self_ptr as *mut u8,
+            Self::layout_for_size((*self_ptr).capacity),
+        )
+    }
 
-    assert_eq!(min_cap, original_capacity_from_repr(1));
-    assert_eq!(min_cap * 2, original_capacity_from_repr(2));
-    assert_eq!(min_cap * 4, original_capacity_from_repr(3));
-    assert_eq!(min_cap * 8, original_capacity_from_repr(4));
-    assert_eq!(min_cap * 16, original_capacity_from_repr(5));
-    assert_eq!(min_cap * 32, original_capacity_from_repr(6));
-    assert_eq!(min_cap * 64, original_capacity_from_repr(7));
+    /// Returns the start of the data section of this buffer which starts behind
+    /// the `Shared` header.
+    ///
+    /// # Safety
+    ///
+    /// This method is only safe to be called on null pointers and pointers
+    /// returned from `Shared::allocate_for_size` which which have not been
+    /// freed using `release_refcount`.
+    unsafe fn data_ptr_mut(&mut self) -> NonNull<u8> {
+        let shared_addr = self as *mut Self;
+        let data_addr = shared_addr.offset(1) as *mut u8;
```

Comment:
FYI, under Stacked Borrows this pointer only has provenance over `Self`, so using it to access the data section would be UB. Under PNVI (the proposed C aliasing model) I believe this is allowed. It's not entirely clear what the finalized rules for Rust will be, but you should probably do one of:

* Fix this under SB, probably by adding a `[u8]` member to `Shared`
* Fix this when it becomes UB for sure
* Campaign for this to not become UB, perhaps on https://github.com/rust-lang/unsafe-code-guidelines/issues/256